src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,741 @@
+        }
+
+        public override void ExtractTraceIdAndState(object? carrier, PropagatorGetterCallback? getter, out string? traceId, out string? traceState)
+        {
+            if (getter is null)
+            {
+                traceId = null;
+                traceState = null;
+                return;
+            }
+
+            getter(carrier, TraceParent, out traceId, out _);
+            if (IsInvalidTraceParent(traceId))
+            {
+                traceId = null;
+            }
+
+            getter(carrier, TraceState, out string? traceStateValue, out _);
+            traceState = ValidateTraceState(traceStateValue);
+        }
+
+        public override IEnumerable<KeyValuePair<string, string?>>? ExtractBaggage(object? carrier, PropagatorGetterCallback? getter)
+        {
+            if (getter is null)
+            {
+                return null;
+            }
+
+            getter(carrier, Baggage, out string? theBaggage, out _);
+
+            TryExtractBaggage(theBaggage, out IEnumerable<KeyValuePair<string, string?>>? baggage);
+
+            return baggage;
+        }
+
+        internal static bool TryExtractBaggage(string? baggageString, out IEnumerable<KeyValuePair<string, string?>>? baggage)
+        {
+            baggage = null;
+            List<KeyValuePair<string, string?>>? baggageList = null;
+
+            if (string.IsNullOrEmpty(baggageString))
+            {
+                return true;
+            }
+
+            ReadOnlySpan<char> baggageSpan = baggageString;
+
+            do
+            {
+                int entrySeparator = baggageSpan.IndexOf(Comma);
+                ReadOnlySpan<char> currentEntry = entrySeparator >= 0 ? baggageSpan.Slice(0, entrySeparator) : baggageSpan;
+
+                int keyValueSeparator = currentEntry.IndexOf(Equal);
+                if (keyValueSeparator <= 0 || keyValueSeparator >= currentEntry.Length - 1)
+                {
+                    break; // invalid format
+                }
+
+                ReadOnlySpan<char> keySpan = currentEntry.Slice(0, keyValueSeparator);
+                ReadOnlySpan<char> valueSpan = currentEntry.Slice(keyValueSeparator + 1);
+
+                if (TryDecodeBaggageKey(keySpan, out string? key) && TryDecodeBaggageValue(valueSpan, out string value))
+                {
+                    baggageList ??= new List<KeyValuePair<string, string?>>();
+                    baggageList.Add(new KeyValuePair<string, string?>(key, value));
+                }
+
+                baggageSpan = entrySeparator >= 0 ? baggageSpan.Slice(entrySeparator + 1) : ReadOnlySpan<char>.Empty;
+            } while (baggageSpan.Length > 0);
+
+            baggage = baggageList;
+            return baggageList != null;
+        }
+
+        // list  = list-member 0*31( OWS "," OWS list-member )
+        // list-member = (key "=" value) / OWS
+        //
+        // key = ( lcalpha / DIGIT ) 0*255 ( keychar )
```

Comment:
is this saying that a key is an lcalpha or digit, up to 255 of them followed by a keychar?